### PR TITLE
[RHPAM-2774] AchievedAtDate for milestone is not preserved when case is reopen

### PR DESCRIPTION
**Pull request by elguardian**
modify query for milestones as a plain value object

### CODE REVIEW

File: jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/CaseRuntimeDataServiceImpl.java [hunk 1]
```diff
@@ -340,43 +340,36 @@ public CaseDefinition getCase(String deploymentId, String caseDefinitionId) {
...
+                                                                                                                                     queryContext);
+
+        // achieved milestones
+        Map<String, List<CaseMilestoneInstance>> milestonesGroup = nodes.stream()
+                                                                        .map(n -> (NodeInstanceDesc) n)
+                                                                        .filter(n -> n.getType() == 1)
+                                                                        .map(n -> new CaseMilestoneInstanceImpl(String.valueOf(n.getId()), n.getName(), true, n.getDataTimeStamp()))
+                                                                        .collect(Collectors.groupingBy(CaseMilestoneInstance::getName));
+
+        List<CaseMilestoneInstance> milestones = milestonesGroup.values().stream().map(e -> e.stream().min((o1, o2) -> o1.getAchievedAt().compareTo(o2.getAchievedAt())).get()).collect(toList());
```
Comment (gmunozfe) [inline jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/CaseRuntimeDataServiceImpl.java]: The map contains lists with one element, so ordering is applying to that list and therefore not working.
Perhaps you can order after generating the list of CaseMilestoneInstance.

`milestones.sort(Comparator.comparing(CaseMilestoneInstance::getAchievedAt));`

Comment (elguardian) [reply jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/CaseRuntimeDataServiceImpl.java]: Not really sure I understand your comments. If a case is cancelled and then reopen it should contain at least two elements and so on ?... do you mind to check again ?

Comment (gmunozfe) [reply jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/CaseRuntimeDataServiceImpl.java]: I meant for the failing test, which contained only one element per key, and the list obtained from the map was not ordered. Now, it's working fine, thanks!

**Comment by gmunozfe** [comment]
Check this scenario:

No milestone achieved before cancelling case. After reopening, milestone is achieved.
When retrieving milestones, however, _achievedAt_ date corresponds to _processInstanceId=1_, when during that instance, no milestone was achieved (it was achieved for _proccessInstanceId=2_).

It seems like that node is updated to _completed=true_, when it shouldn't:

_NodeInstanceDesc{id=1, nodeId=PID__d78219e8-fdde-4c24-9ed6-8864d7b2d26b, nodeUniqueId=PID__d78219e8-fdde-4c24-9ed6-8864d7b2d26b, name=All work done, deploymentId=org.jbpm.cases:case-module:1.0.0, **processInstanceId=1**, type=MilestoneNode, **completed=true**, dataTimeStamp=2020-03-09 18:04:02.748}]_

**Comment by elguardian** [comment]
you were right @gmunozfe but there was not a way to know the type of node exit. I just added the flat to the node instance log to be able to discriminate between them. This should work.

**Comment by gmunozfe** [comment]
works fine, thanks @elguardian, awesome job!

**Comment by elguardian** [comment]
@gmunozfe I need you check the scripting node instance log new column and feedback

File: jbpm-installer/src/test/resources/ddl60/postgresql/postgresql-jbpm-schema.sql [hunk 2]
```diff
@@ -130,6 +130,7 @@
         processInstanceId int8 not null,
         type int4 not null,
         workItemId int8,
+        skipable boolean,
```
Comment (gmunozfe) [inline jbpm-installer/src/test/resources/ddl60/postgresql/postgresql-jbpm-schema.sql]: shouldn't it be "cancel"?

File: jbpm-installer/src/test/resources/ddl60/db2/db2-jbpm-schema.sql [hunk 3]
```diff
@@ -130,6 +130,7 @@
         processInstanceId bigint not null,
         type integer not null,
         workItemId bigint,
+        cancel smallint not null,
```
Comment (gmunozfe) [inline jbpm-installer/src/test/resources/ddl60/db2/db2-jbpm-schema.sql]: do we need to update ddl60 schemas? my understanding is that this is used to test migration from 6.0 to 7.x

File: jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/CaseRuntimeDataServiceImpl.java [hunk 4]
```diff
@@ -340,44 +340,47 @@ public CaseDefinition getCase(String deploymentId, String caseDefinitionId) {
...
-        
-        return applyPagination(milestones, queryContext);
+
+        List<CaseMilestoneInstance> sortedMilestones = milestones.stream().sorted(this::compareMilestones).collect(toList());
+        return applyPagination(sortedMilestones, queryContext);
+    }
+
+    private int compareMilestones(CaseMilestoneInstance o1, CaseMilestoneInstance o2) {
+        if (o1.getAchievedAt() == null) {
+            return -1;
```
Comment (gmunozfe) [inline jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/CaseRuntimeDataServiceImpl.java]: if both are null, shouldn't it return 0?

Comment (elguardian) [reply jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/CaseRuntimeDataServiceImpl.java]: null <> null = true... null == null = false... depends on our view.

File: jbpm-installer/src/test/resources/ddl60/db2/db2-jbpm-schema.sql (hunk 3 above)
Comment (elguardian) [reply jbpm-installer/src/test/resources/ddl60/db2/db2-jbpm-schema.sql]: this flag will be changed in exchange for new type of nodeinstancelog

**Comment by elguardian** [comment]
depends on https://github.com/kiegroup/droolsjbpm-knowledge/pull/431

**Pull request merged**